src/compiler/csharp_generator.cc

```diff
@@ -289,114 +280,126 @@ std::string GetMethodReturnTypeServer(const MethodDescriptor *method) {
       return "";
     case METHODTYPE_SERVER_STREAMING:
     case METHODTYPE_BIDI_STREAMING:
       return ", grpc::IServerStreamWriter<" +
-             GetClassName(method->output_type()) + "> responseStream";
+             method->output_type_name() + "> responseStream";
   }
   GOOGLE_LOG(FATAL) << "Can't get here.";
   return "";
 }
 
 // Gets vector of all messages used as input or output types.
-std::vector<const Descriptor *> GetUsedMessages(
-    const ServiceDescriptor *service) {
-  std::set<const Descriptor *> descriptor_set;
-  std::vector<const Descriptor *>
-      result;  // vector is to maintain stable ordering
+std::vector<grpc::string> GetUsedMessages(
+      const grpc_generator::Service *service) {
+  std::set<grpc::string> descriptor_set;
+  std::vector<grpc::string> result;  // vector is to maintain stable ordering
   for (int i = 0; i < service->method_count(); i++) {
-    const MethodDescriptor *method = service->method(i);
-    if (descriptor_set.find(method->input_type()) == descriptor_set.end()) {
-      descriptor_set.insert(method->input_type());
-      result.push_back(method->input_type());
+    const grpc_generator::Method *method = service->method(i).get();
+    if (descriptor_set.find(method->input_type_name()) == descriptor_set.end()) {
+      descriptor_set.insert(method->input_type_name());
+      result.push_back(method->input_type_name());
     }
-    if (descriptor_set.find(method->output_type()) == descriptor_set.end()) {
-      descriptor_set.insert(method->output_type());
-      result.push_back(method->output_type());
+    if (descriptor_set.find(method->output_type_name()) == descriptor_set.end()) {
+      descriptor_set.insert(method->output_type_name());
+      result.push_back(method->output_type_name());
     }
   }
   return result;
 }
 
-void GenerateMarshallerFields(Printer *out, const ServiceDescriptor *service) {
-  std::vector<const Descriptor *> used_messages = GetUsedMessages(service);
+
+void GenerateMarshallerFields(Printer *out, const grpc_generator::Service *service) {
+  std::vector<grpc::string> used_messages = GetUsedMessages(service);
   for (size_t i = 0; i < used_messages.size(); i++) {
-    const Descriptor *message = used_messages[i];
+    grpc::string message = used_messages[i];
+    std::map<grpc::string, grpc::string> vars;
+    vars["fieldname"] = GetMarshallerFieldName(message);
+    vars["type"] = message;
     out->Print(
-        "static readonly grpc::Marshaller<$type$> $fieldname$ = "
-        "grpc::Marshallers.Create((arg) => "
-        "global::Google.Protobuf.MessageExtensions.ToByteArray(arg), "
-        "$type$.Parser.ParseFrom);\n",
-        "fieldname", GetMarshallerFieldName(message), "type",
-        GetClassName(message));
+      vars,
+      "static readonly grpc::Marshaller<$type$> $fieldname$ = "
+      "grpc::Marshallers.Create((arg) => "
+      "global::Google.Protobuf.MessageExtensions.ToByteArray(arg), " //ERR ref to pb
```

Comment:
You'll need to replace this code with a C# interface that we implement for protobuf to call this particular function. We'll then later also implement it for FlatBuffers, and since in FlatBuffers it is already a byte-array, it will be a no-op function. Important to design the interface such that in FlatBuffers it does not require an unnecessary copy.

Reply:
Currently I don't have a good answer here - hence my concern if the generated code is reusable like this.
Looks like we are trying to impose the code structure (that was designed for protobufs) on other serialization formats (such as flattbuffers) and I'm not convinced that's going to work well (and if its going to be maintainable).